message/messages.go

```diff
@@ -186,69 +214,116 @@ func (mb *msgBuilder) marshal(
 	//
 	// This recursive packing allows us to avoid an extra compression on/off
 	// field in the message.
-	startTime := time.Now()
-	compressedBytes, err := mb.gzipCompressor.Compress(uncompressedMsgBytes)
-	if err != nil {
-		return nil, 0, 0, err
+	var (
+		startTime               = time.Now()
+		compressedMsg           p2p.Message
+		opToCompressTimeMetrics map[Op]metric.Averager
+	)
+	switch compressionType {
+	case compression.TypeNone:
+		return uncompressedMsgBytes, 0, op, nil
+	case compression.TypeGzip:
+		compressedBytes, err := mb.gzipCompressor.Compress(uncompressedMsgBytes)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		compressedMsg = p2p.Message{
+			Message: &p2p.Message_CompressedGzip{
+				CompressedGzip: compressedBytes,
+			},
+		}
+		opToCompressTimeMetrics = mb.gzipCompressTimeMetrics
+	case compression.TypeZstd:
+		compressedBytes, err := mb.zstdCompressor.Compress(uncompressedMsgBytes)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		compressedMsg = p2p.Message{
+			Message: &p2p.Message_CompressedZstd{
+				CompressedZstd: compressedBytes,
+			},
+		}
+		opToCompressTimeMetrics = mb.zstdCompressTimeMetrics
+	default:
+		return nil, 0, 0, errUnknownCompressionType
 	}
 
-	compressedMsg := p2p.Message{
-		Message: &p2p.Message_CompressedGzip{
-			CompressedGzip: compressedBytes,
-		},
-	}
 	compressedMsgBytes, err := proto.Marshal(&compressedMsg)
 	if err != nil {
 		return nil, 0, 0, err
 	}
 	compressTook := time.Since(startTime)
 
+	if compressTimeMetric, ok := opToCompressTimeMetrics[op]; ok {
+		// This case should always execute, but check just in case.
```

Comment:
Is it worth having a logger in this struct to log a warning here?

Reply:
previously the node would just crash... -shrug-

Reply:
added a log